### PR TITLE
added test run time limits

### LINKED ISSUE #889: Limit Duration of Each Individual Test
It would be useful to be able to limit each test to a number of seconds.

### PR DESCRIPTION
**Pull request by BebeSparkelSparkel**
Allow tests to have limited run time via a command line argument or Spec modifiers.

closes #889

### CODE REVIEW

File: hspec-core/src/Test/Hspec/Core/Config/Definition.hs
Comment (sol) [inline hspec-core/src/Test/Hspec/Core/Config/Definition.hs:361]: What do you think about naming the option `--timeout` instead of `--max-time-per-test`? Would that work for you? 

I think we also want `--no-timeout` to set it back to `Nothing`. Config options [can come from various places](https://hspec.github.io/options.html#precedence-resolution-for-options-that-are-specified-in-multiple-places), so you may want to set it back to nothing at the command line.

**Review by sol — CHANGES_REQUESTED** [review]
This looks great!

- [ ] Please split this into two separate PRs, that is move everything "per item" into a new "draft" PR (or keep them locally for now, whatever works best for you).  Let's then get the command line option merged first and do "per item" as a follow up.
- [ ] This needs a tests! Maybe just a single acceptance test will be fine, look at https://github.com/hspec/hspec/blob/main/hspec-core/test/Test/Hspec/Core/RunnerSpec.hs

NB: Things like timeouts are tricky  to test, as we don't want to unnecessarily increase the runtime of the test suit. Still, maybe a timeout value of zero will do the trick, or maybe we will remove the test again later.

**Comment by BebeSparkelSparkel** [comment]
I added tests but am using `hspecWithResult` which mucks up the test results. Is there a function that does not produce output or a function that suppresses the output?

**Comment by sol** [comment]
> I am unsure how my changes caused _api-check / api-check (pull_request)_ to fail

This is due to GHC 9.10.1.  I fixed that with #892, please rebase once this is on `main`.

I also added a section to the README on how to update API dumps.  You will need to run that. The purpose of keeping track of API changes explicitly is so that we don't  expose  stuff that we didn't meant to, and to make it easier to determine whether we need a major or minor version bump.

File: hspec-core/src/Test/Hspec/Core/Runner.hs
Comment (sol) [reply hspec-core/src/Test/Hspec/Core/Runner.hs:460]: I think that's due to GHC 9.10.1, should be fixed now.

File: hspec-core/src/Test/Hspec/Core/Clock.hs
Comment (sol) [inline hspec-core/src/Test/Hspec/Core/Clock.hs:57]: Is it possible to implement this in terms of `measure`?

```haskell
measureWithTimeout maxTime action = measure $ maybe (fmap Just) timeout maxTime action
```

File: hspec-core/src/Test/Hspec/Core/Config/Definition.hs
Comment (sol) [inline hspec-core/src/Test/Hspec/Core/Config/Definition.hs:94]: Please move this closer to related options, e.g. after `configFailOnEmptyDescription` (matching the location in the help message).

File: hspec-core/src/Test/Hspec/Core/Config/Definition.hs
Comment (sol) [inline hspec-core/src/Test/Hspec/Core/Config/Definition.hs:142]: Same here

File: hspec-core/test/Test/Hspec/Core/RunnerSpec.hs
Comment (sol) [inline hspec-core/test/Test/Hspec/Core/RunnerSpec.hs:1025]: Please move this test case closer to tests for related options, e.g. around line 389.

File: hspec-core/src/Test/Hspec/Core/Runner.hs
Comment (sol) [inline hspec-core/src/Test/Hspec/Core/Runner.hs:242]: I think this is not strictly needed and can be deferred to #891.

File: hspec-core/src/Test/Hspec/Core/Runner.hs
Comment (sol) [inline hspec-core/src/Test/Hspec/Core/Runner.hs:460]: As I understand it, you could use  `maxTime` instead of `configMaxTimePerTest config` here, for now.

Also, that line of code is getting quite long. I think it could make sense to extract some of the code into a separate function, say `applyTimeout` or `failOnTimeout` or both, not sure exactly how it will turn out nice, maybe just try something.

File: hspec-core/src/Test/Hspec/Core/Tree.hs
Comment (sol) [inline hspec-core/src/Test/Hspec/Core/Tree.hs:146]: Same here, I think this can be deferred to #891.

File: hspec-core/src/Test/Hspec/Core/Config/Definition.hs
Comment (sol) [reply hspec-core/src/Test/Hspec/Core/Config/Definition.hs:94]: I try to keep names of command line options and the corresponding fields of the config data type consistent.  So this should be `configTimeout`, or if this is too generic then alternatively `configSpecItemTimeout`. I would probably go with the first.

**Review by sol — CHANGES_REQUESTED** [review]
Hey, this looks great. I left a couple of more comments. I haven't looked at a test case in detail yet, but from a quick glimpse, I assume this doesn't add any delays to the test suite, which is nice. Please  take a look at my comments and make sure CI is green, then I will take a final look.

**Comment by BebeSparkelSparkel** [comment]
I'm not sure how to accommodate the unresolved. Also, if #891 is to be pulled then not storing the timeout in Item is extra work that will be undone.

**Comment by sol** [comment]
@BebeSparkelSparkel you marked a couple of my comments as resolved, but I don't see any associated code changes. Did you forget to push?

**Comment by BebeSparkelSparkel** [comment]
Changed the wrong branch. I'll try and git fix it.

File: hspec-core/src/Test/Hspec/Core/Runner.hs
Comment (BebeSparkelSparkel) [reply hspec-core/src/Test/Hspec/Core/Runner.hs]: > As I understand it, you could use  maxTime instead of configMaxTimePerTest config here, for now.

Sorry, I don't understand this.

File: hspec-core/src/Test/Hspec/Core/Tree.hs
Comment (BebeSparkelSparkel) [reply hspec-core/src/Test/Hspec/Core/Tree.hs]: I am unsure where to get the timeout duration from if this is removed.

**Comment by sol** [comment]
I finalized an extension API that allows additional functionality to be implemented outside of `hspec-core`: https://github.com/hspec/hspec/pull/868

Originally, I didn't want to delay your feature until the API is finalized, but now that I consider it done I think the way forward is to implement this feature as a separate package. This also means that you have complete freedom on how to name things, coding style, etc + you are not blocked by waiting for a code review from me.

@BebeSparkelSparkel there are two options here, either you implement your feature with the git version of `hspec-core`.  This will give you the opportunity to  provide  feedback before the API is released. Or alternatively, I can move forward with a release and you can then use the version from Hackage.  Please let me know what you prefer.

Once your package is on Hackage, I will then add it to Hspec's documentation, so that it can be used by others.

**Comment by sol** [comment]
@BebeSparkelSparkel the PR description contains an example on how to use the API + here is an other example: https://github.com/hspec/hspec/blob/ec2c16712ef68ba41d6f721d32840afadcba9576/hspec-core/src/Test/Hspec/Tags.hs

**Comment by BebeSparkelSparkel** [comment]
I will try the git version implementation and add comment to the PR if I have issues.

Since you were considering adding this to core, should I make this a new package in this repo since this is a fundamental part of testing?

**Comment by sol** [comment]
> I will try the git version implementation and add comment to the PR if I have issues.
> 
> Since you were considering adding this to core, should I make this a new package in this repo since this is a fundamental part of testing?

I suggest a separate package in a separate repo.  We can always decide to upstream things later.  I could imagine to upstream `--timeout` eventually, but maybe not per item timeouts, they could remain in a separate package and I imagine we could still make both of them  interact  nicely with each other. But let's save this discussion for another day.

By the way, if you think a separate repo is too much overhead, you could add it to `hspec-contrib`. I still think a separate package will  provide more visibility. `hspec-contrib` is just a dumping ground for random stuff.

**Comment by BebeSparkelSparkel** [comment]
Ok, do extensions also allow for the implementation of `limit` and `noLimit` that can override the the global ``--timeout`?

**Comment by sol** [comment]
> Ok, do extensions also allow for the implementation of `limit` and `noLimit` that can override the the global `--timeout`?

Yes, I think so. Look at the "tags" example.

**Comment by sol** [comment]
> > Ok, do extensions also allow for the implementation of `limit` and `noLimit` that can override the the global `--timeout`?
> 
> Yes, I think so. Look at the "tags" example.

@BebeSparkelSparkel I think even the "ci" example should demonstrate all the ideas, but it's much shorter. So maybe that's a better place to start. Ask if you need any specific advice or help.

**Comment by BebeSparkelSparkel** [comment]
Thank you I will attempt this now

**Pull request closed without merging**
